File: rn_app/src/lib/calculations.ts
import { roundToNearestFive } from './utils';

/**
 * Calculate target weight based on 1RM and percentage
 * @param oneRepMax - User's current 1RM in lbs
 * @param percentage - Decimal percentage (e.g., 0.65 for 65%)
 * @returns Target weight rounded to nearest 5 lbs
 */
export function calculateTargetWeight(oneRepMax: number, percentage: number): number {
  const rawWeight = oneRepMax * percentage;
  return roundToNearestFive(rawWeight);
}

/**
 * Session result classification for recommendation engine
 */
export type SessionResult = 'success' | 'failure' | 'first';

/**
 * Get next recommended working weight based on session result
 * - 'first': 70% of 1RM (starting weight for a new exercise)
 * - 'success': +5 lbs from current weight
 * - 'failure': -5 lbs from current weight, with a floor at 50% of 1RM
 * All results are rounded to nearest 5 lbs.
 *
 * @param currentWeight - Current working weight in lbs
 * @param result - Outcome of the most recent session
 * @param oneRepMax - User's 1RM for this exercise
 * @returns Recommended weight for next session, rounded to nearest 5 lbs
 */
export function getNextRecommendedWeight(
  currentWeight: number,
  result: SessionResult,
  oneRepMax: number
): number {
  switch (result) {
    case 'first':
      return roundToNearestFive(oneRepMax * 0.7);
    case 'success':
      return roundToNearestFive(currentWeight + 5);
    case 'failure': {
      const floor = roundToNearestFive(oneRepMax * 0.5);
      return Math.max(roundToNearestFive(currentWeight - 5), floor);
    }
  }
}

/**
 * Determine if a single set was successfully completed
 * A set is successful if actualReps >= prescribedReps AND
 * (if prescribedWeight is defined) actualWeight >= prescribedWeight.
 */
export function wasSetSuccessful(set: {
  actualReps: number;
  prescribedReps: number;
  actualWeight: number;
  prescribedWeight?: number;
}): boolean {
  const repsOk = set.actualReps >= set.prescribedReps;
  const weightOk =
    set.prescribedWeight === undefined || set.actualWeight >= set.prescribedWeight;
  return repsOk && weightOk;
}

/**
 * Determine if an entire session (array of sets) was successfully completed
 * A session is successful if ALL sets pass wasSetSuccessful.
 * An empty array returns true (vacuously successful).
 */
export function wasSessionSuccessful(
  sets: Array<{
    actualReps: number;
    prescribedReps: number;
    actualWeight: number;
    prescribedWeight?: number;
  }>
): boolean {
  return sets.every(wasSetSuccessful);
}

/**
 * Check if a weight represents a new personal record
 * @param weight - Current weight
 * @param previousMax - Previous maximum weight
 * @returns True if weight exceeds previous max
 */
export function isPersonalRecord(weight: number, previousMax: number): boolean {
  return weight > previousMax;
}

/**
 * Calculate volume load (weight × reps × sets)
 * @param weight - Weight lifted
 * @param reps - Reps completed
 * @param sets - Number of sets
 * @returns Total volume load
 */
export function calculateVolumeLoad(weight: number, reps: number, sets: number): number {
  return weight * reps * sets;
}

/**
 * Determine if user has plateaued (no progress in 3+ workouts)
 * @param weights - Array of weights from recent workouts, ordered by date
 * @returns True if plateau detected
 */
export function detectPlateau(weights: number[]): boolean {
  if (weights.length < 3) return false;

  const lastThree = weights.slice(-3);
  const maxWeight = Math.max(...lastThree);
  const minWeight = Math.min(...lastThree);

  // Plateau if variance is less than 5 lbs
  return maxWeight - minWeight < 5;
}

File: flutter_app/lib/core/recommendations/calculations.dart
/// Pure Dart recommendation calculations.
/// Exact port of v1.1 src/lib/calculations.ts.
///
/// All functions are pure (no side effects, no DB access).
/// Used by plateau_detection.dart and coaching UI.

/// Round to nearest 5 lbs.
double roundToNearestFive(double value) {
  return (value / 5).round() * 5.0;
}

/// Epley formula for estimated 1RM.
/// Caps reps at 10 (higher reps are unreliable for 1RM estimation).
/// Matches v1.1: `weight * (1 + Math.min(reps, 10) / 30)`.
double epley(double weight, int reps) {
  if (reps <= 1) return weight;
  return weight * (1 + reps.clamp(1, 10) / 30);
}

/// Session result classification for recommendation engine.
enum SessionResult { success, failure, first }

/// Get next recommended working weight based on session result.
/// - first:   70% of 1RM (starting weight for a new exercise)
/// - success: +5 lbs from current weight
/// - failure: -5 lbs from current weight, floor at 50% of 1RM
/// All results are rounded to nearest 5 lbs.
double getNextRecommendedWeight(
  double currentWeight,
  SessionResult result,
  double oneRepMax,
) {
  switch (result) {
    case SessionResult.first:
      return roundToNearestFive(oneRepMax * 0.7);
    case SessionResult.success:
      return roundToNearestFive(currentWeight + 5);
    case SessionResult.failure:
      final floor = roundToNearestFive(oneRepMax * 0.5);
      return (roundToNearestFive(currentWeight - 5)).clamp(floor, double.infinity);
  }
}

/// Check if a single set was successfully completed.
/// A set is successful if actualReps >= prescribedReps AND
/// (if prescribedWeight is defined) actualWeight >= prescribedWeight.
bool wasSetSuccessful({
  required int actualReps,
  required int prescribedReps,
  required double actualWeight,
  double? prescribedWeight,
}) {
  final repsOk = actualReps >= prescribedReps;
  final weightOk = prescribedWeight == null || actualWeight >= prescribedWeight;
  return repsOk && weightOk;
}

/// Check if all sets in a session were successfully completed.
/// An empty list returns true (vacuously successful — matches v1.1).
bool wasSessionSuccessful(List<Map<String, dynamic>> sets) {
  return sets.every(
    (s) => wasSetSuccessful(
      actualReps: s['actualReps'] as int,
      prescribedReps: s['prescribedReps'] as int,
      actualWeight: s['actualWeight'] as double,
      prescribedWeight: s['prescribedWeight'] as double?,
    ),
  );
}

/// Check if a weight represents a new personal record.
bool isPersonalRecord(double weight, double previousMax) {
  return weight > previousMax;
}

/// Calculate volume load: weight × reps × sets.
double calculateVolumeLoad(double weight, int reps, int sets) {
  return weight * reps * sets;
}

/// Detect plateau: variance < 5 lbs in last 3 weights.
/// Returns false if fewer than 3 weights provided.
bool detectPlateau(List<double> weights) {
  if (weights.length < 3) return false;
  final lastThree = weights.sublist(weights.length - 3);
  final maxWeight = lastThree.reduce((a, b) => a > b ? a : b);
  final minWeight = lastThree.reduce((a, b) => a < b ? a : b);
  return maxWeight - minWeight < 5;
}
